### PR TITLE
Implement `script.addPreloadScript`

### PR DESCRIPTION
**Pull request by thiagowfx**
Bug: #293
CDP: https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-addScriptToEvaluateOnNewDocument
Spec: https://w3c.github.io/webdriver-bidi/#command-script-addPreloadScript

### CODE REVIEW

File: src/bidiMapper/domains/context/browsingContextProcessor.ts [hunk 1]
```diff
@@ -291,14 +291,29 @@ export class BrowsingContextProcessor {
...
 
-  // eslint-disable-next-line @typescript-eslint/require-await
   async process_script_addPreloadScript(
-    _params: Script.AddPreloadScriptParameters
+    params: Script.AddPreloadScriptParameters
   ): Promise<Script.AddPreloadScriptResult> {
-    throw new Message.UnknownErrorException('Not implemented.');
+    const cdpClient = this.#cdpConnection.browserClient();
+
+    const result1 = await cdpClient.sendCommand('Runtime.callFunctionOn', {
```
Comment (sadym-chromium) [inline src/bidiMapper/domains/context/browsingContextProcessor.ts:299]: [`Page.addScriptToEvaluateOnNewDocument`](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-addScriptToEvaluateOnNewDocument) is the right CDP method.
```suggestion
    const result1 = await cdpClient.sendCommand('Page.addScriptToEvaluateOnNewDocument', {
```

Comment (sadym-chromium) [inline src/bidiMapper/domains/context/browsingContextProcessor.ts:299]: This line is not needed. On the browser client you don't have a runtime.

Comment (thiagowfx) [reply src/bidiMapper/domains/context/browsingContextProcessor.ts:299]: Ack, removed.

File: src/bidiMapper/domains/context/browsingContextProcessor.ts [hunk 2]
```diff
@@ -291,14 +291,29 @@ export class BrowsingContextProcessor {
...
+      functionDeclaration: `function() {
+          const script = document.createElement('script');
+          script.src = '${params.expression}';
+          document.documentElement.appendChild(script);
+          }`,
+      // executionContextId: realm.executionContextId,
+    });
+
+    const result2 = await cdpClient.sendCommand(
```
[the remark this answers is not in the document]
Comment (thiagowfx) [reply src/bidiMapper/domains/context/browsingContextProcessor.ts]: There's no browsing context param. Should I get the first element of `getTopLevelContexts`?

[the remark this answers is not in the document]
Comment (thiagowfx) [reply src/bidiMapper/domains/context/browsingContextProcessor.ts]: > So you need to extend the Script.AddPreloadScriptParameters with context for debugging purposes.

That is not in the spec though, should we change the spec to add the context to it?

[the remark this answers is not in the document]
Comment (thiagowfx) [reply src/bidiMapper/domains/context/browsingContextProcessor.ts]: Done, let me know if it looks like what you envisioned.

File: src/bidiMapper/domains/context/browsingContextImpl.ts [hunk 3]
```diff
@@ -530,4 +530,19 @@ export class BrowsingContextImpl {
...
     };
   }
+
+  async addPreloadScript(
+    params: Script.AddPreloadScriptParameters
+  ): Promise<Script.AddPreloadScriptResult> {
+    const result = await this.#cdpTarget.cdpClient.sendCommand(
+      'Page.addScriptToEvaluateOnNewDocument',
+      {
+        source: params.expression,
```
Comment (thiagowfx) [inline src/bidiMapper/domains/context/browsingContextImpl.ts:540]: Need to use `params.sandbox` somewhere.

Would it map to `worldName` in the CDP param?

File: src/bidiMapper/domains/context/browsingContextProcessor.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply src/bidiMapper/domains/context/browsingContextProcessor.ts]: > > So you need to extend the Script.AddPreloadScriptParameters with context for debugging purposes.
> 
> That is not in the spec though, should we change the spec to add the context to it?

For now you can add the parameter to the command parameters, and remove it later on, when you implement the proper handling all the browsing contexts.

Also there is an issue https://github.com/w3c/webdriver-bidi/issues/373 to add the optional field into the command, and you can add spec for the param as well.

File: src/bidiMapper/domains/context/browsingContextImpl.ts (hunk 3 above)
Comment (sadym-chromium) [reply src/bidiMapper/domains/context/browsingContextImpl.ts:540]: yes

**Comment by sadym-chromium** [comment]
Nice! Next steps:
1. Add e2e steps.
2. Make `context` param optional. In such a case you have to track all the new Browsing contexts, and add the `Page.addScriptToEvaluateOnNewDocument` for each of them.
3. Implement `script.removePreloadScript`

**Comment by thiagowfx** [comment]
> Implement script.removePreloadScript.

I'll start with this one as it's more straightforward

File: src/bidiMapper/domains/context/browsingContextImpl.ts (hunk 3 above)
Comment (thiagowfx) [reply src/bidiMapper/domains/context/browsingContextImpl.ts:540]: Done

**Comment by thiagowfx** [comment]
Basic E2E test added

File: src/bidiMapper/domains/context/browsingContextProcessor.ts [hunk 4]
```diff
@@ -291,15 +291,29 @@ export class BrowsingContextProcessor {
...
-    throw new Message.UnknownErrorException('Not implemented.');
+    const contexts: BrowsingContextImpl[] = [];
+    const scripts: Script.AddPreloadScriptResult[] = [];
 
-    return {
-      script: '', // TODO(#293): Populate script.
-    };
+    if (params.context) {
+      contexts.push(
+        this.#browsingContextStorage.getKnownContext(params.context)
```
Comment (sadym-chromium) [inline src/bidiMapper/domains/context/browsingContextProcessor.ts:302]: Here is one edge case with OOPiF. When a frame is moved out of process, we have to add those scripts as well. Can be covered later, but I want to make sure we don't forget about it.

File: tests/test_preload_script.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                "expression": "() => { window.foo='bar'; }",
+                "context": context_id,
+            }
+        })
+
+    resp = await read_JSON_message(websocket)
+
+    assert resp == {'id': 1, 'script': '1'}
+
+    result = await execute_command(
```
Comment (sadym-chromium) [inline tests/test_preload_script.py:37]: You should navigate somewhere first so that a new frame is created.

File: src/protocol/protocol.ts [hunk 6]
```diff
@@ -693,6 +693,7 @@ export namespace Script {
   export type AddPreloadScriptParameters = {
     expression: string;
     sandbox?: string;
+    context?: CommonDataTypes.BrowsingContext;
   };
 
   export type AddPreloadScriptResult = {
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply src/protocol/protocol.ts:699]: Meaning it should look like 
```
  export type AddPreloadScriptResult = {
    result: {
      script: PreloadScript;
    };
  };
```

File: src/bidiMapper/domains/context/browsingContextImpl.ts [hunk 7]
```diff
@@ -530,4 +530,20 @@ export class BrowsingContextImpl {
...
```
Comment (sadym-chromium) [inline src/bidiMapper/domains/context/browsingContextImpl.ts:540]: `params.expression` is a function which has to be called (with bindings as arguments, but it's for later), while `Page.addScriptToEvaluateOnNewDocument. source` is a string literal (callFunction vs evaluate).

File: tests/test_preload_script.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        'type': "success",
+        "result": {
+            "type": "string",
+            "value": 'bar'
+        }
+    }
+
+
+@pytest.mark.asyncio
+async def test_addPreloadScript_logging(websocket, context_id):
```
Comment (sadym-chromium) [inline tests/test_preload_script.py:60]: This test should:
0. Subscribe to `console.log` event.
1. `execute_command(script.addPreloadScript)`.
2. `send_JSON_command(navigate(..., 'complete'))`.
3. `read_JSON_message()` and assert it is a log message.
4. `read_JSON_message()` and assert it is a navigation result.

**Comment by sadym-chromium** [comment]
Pleas note that most of my comments are fixed in https://github.com/GoogleChromeLabs/chromium-bidi/pull/586

File: tests/test_preload_script.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                "expression": "() => console.log('my preload script');",
+                "context": context_id,
+            }
+        })
+
+    resp = await read_JSON_message(websocket)
+
+    assert resp == {'id': 1, 'script': '1'}
+
+    await wait_for_event(websocket, "log.entryAdded")
```
[the remark this answers is not in the document]
Comment (thiagowfx) [reply tests/test_preload_script.py]: Ack

**Comment by thiagowfx** [comment]
Added TODOs for missing edge cases.
